### PR TITLE
Contributor project automations via workflows

### LINKED ISSUE #960: GitHub Actions: Can PR authors be given the ability to move cards between Project columns?
We use a GitHub Project named [Code Reviews](https://github.com/microsoft/STL/projects/1) to organize our work. It has several columns, including Work In Progress and Ready For Initial Review. We've enabled automation to make life easier (e.g. requesting changes will automatically move a PR to the Work In Progress column, which is amazing).

We want to give PR authors some control over the placement of their PRs in this Project. Specifically, it would be great if a PR author could move their own PR to either the Work In Progress or Ready For Initial Review column, regardless of its current location. (Because PR authors can push changes to their own branches, this indicates whether they're actively working on a PR, or whether they're waiting for maintainer review.)

GitHub Actions/Workflows may be capable of this. #829 (reverted by #847) was doomed because PRs from forks (i.e. all of our PRs) don't have access to secrets in the base repo. However, perhaps there could be a workaround, like a PR author could comment in an issue "/wip GH-nnn" or "/review GH-nnn" and a workflow would compare the commenter and the PR's submitter, and move the card if they match. I don't know if that would be doomed too, but presumably there would be no security concerns for a workflow running on issue comments (since only the base repo is involved there).

Suggested by @miscco in https://github.com/microsoft/STL/issues/921#issuecomment-650523097.

### PR DESCRIPTION
**Pull request by sam20908**
Fixes #960

This PR adds automation for contributors to move their PR between "Work In Progress" and "Initial Review".

Previous attempt by https://github.com/microsoft/STL/pull/829 was very close, but it needed `pull_request_target` to allow fork PRs to access secrets. This technically builds upon that (even I didn't know there was a previous attempt when I implemented this).

#### Automation So Far
- `/pr wip` moves the PR card to "Work In Progress". It only searches through "Initial Review" and "Final Review" for now (because the assumption right now is that PRs that got through those reviews are ready to go), can change if requested.
- `/pr review` moves the PR card to "Initial Review", it only searches through "Work In Progress" for the PR card.

#### Other Automation Ideas
- Automatically adds PR as a card to "Initial Review" in "Code Reviews" project when the PR is submitted
- Drafted PRs automatically puts the PR into "work in progress" status, same effect as `/pr wip`

#### Design Comments
- No idea if a secret is really needed for STL (at least I didn't, tested with a separate account to fork the experimental repository and created a PR).
- Moving the card requires JavaScript as there is no capable Actions to handle the way its being triggered, although there is https://github.com/alex-page/github-project-automation-plus, but it requires getting triggered through `issues`, and another workflow adding label wouldn't trigger it by default.
- There was no API to get the project and card the PR is assigned to, so majority of moving card code was to find out where the corresponding card is related through API calls.

Experimentation repository can be found here: https://github.com/sam20908/STL-PR-Workflow

### CODE REVIEW

File: .github/workflows/move-ready-for-review-prs.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+            // Since .then is asynchronous, cannot guarantee work_in_progress is set before
+            // the PR card is found. Just loop through the columns again
+            columns.data.forEach(column => {
+              if (column.name === 'Initial Review') {
+                work_in_progress = column.id
+              }
+            })
+
+            // FIXME figure out best way to handle error
```
Comment (sam20908) [inline .github/workflows/move-ready-for-review-prs.yml:52]: Reminder.

File: .github/workflows/move-ready-for-review-prs.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            columns.data.forEach(column => {
+              github.rest.projects.listCards({
+                column_id: column.id,
+              }).then(cards => {
+                  cards.data.forEach(card => {
+                    if (card.content_url === context.payload.issue.url) {
+                      pr_card = card.id // TODO break out early somehow, avoid extra work
```
Comment (sam20908) [inline .github/workflows/move-ready-for-review-prs.yml:59]: Given the code here is executed asynchronously, is this possible?

File: .github/workflows/move-ready-for-review-prs.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            // Also find the card of the PR in all columns
+            // This avoids hardcoding the column ID and card ID
+            const columns = await github.rest.projects.listColumns({
+              project_id: code_reviews.id,
+            });
+
+            let work_in_progress = null
+            let pr_card = null
+            // FIXME forEach is needed instead of a for loop, otherwise it doesn't work
+            // FIXME find out why
```
Comment (sam20908) [inline .github/workflows/move-ready-for-review-prs.yml:42]: Reminder.

**Review by sam20908 — COMMENTED** [review]
Some of the JavaScript questions so far.

File: .github/workflows/move-ready-for-review-prs.yml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            }
+
+            // Find "Initial Review" column manually by name matching
+            // Also find the card of the PR in all columns
+            // This avoids hardcoding the column ID and card ID
+            const columns = await github.rest.projects.listColumns({
+              project_id: code_reviews.id,
+            });
+
+            let work_in_progress = null
```
Comment (miscco) [inline .github/workflows/move-ready-for-review-prs.yml:39]: There is a array.find method you should be able to use
```
const wip = Columbus.data.find(column => column.name === 'Initial Review');
if (!wip) {
    return;
}

```

File: .github/workflows/move-ready-for-review-prs.yml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            let pr_card = null
+            // FIXME forEach is needed instead of a for loop, otherwise it doesn't work
+            // FIXME find out why
+
+            // Since .then is asynchronous, cannot guarantee work_in_progress is set before
+            // the PR card is found. Just loop through the columns again
+            columns.data.forEach(column => {
+              if (column.name === 'Initial Review') {
```
Comment (miscco) [inline .github/workflows/move-ready-for-review-prs.yml:47]: I am confused why that looks for `Initial Reviw`

File: .github/workflows/move-ready-for-review-prs.yml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                work_in_progress = column.id
+              }
+            })
+
+            // FIXME figure out best way to handle error
+            columns.data.forEach(column => {
```
Comment (miscco) [inline .github/workflows/move-ready-for-review-prs.yml:53]: This looks for All possible columns, but we should only move from `work in progress` to `Initial review`

File: .github/workflows/move-ready-for-review-prs.yml [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+              github.rest.projects.listCards({
+                column_id: column.id,
+              }).then(cards => {
+                  cards.data.forEach(card => {
+                    if (card.content_url === context.payload.issue.url) {
+                      pr_card = card.id // TODO break out early somehow, avoid extra work
+
+                      if (work_in_progress !== null) {
```
Comment (miscco) [inline .github/workflows/move-ready-for-review-prs.yml:61]: I think overall the code should look like something along those lines
```typescript
const initial_review = columns.data.find(column => column.name === 'Initial Review');
if (!initial_review) {
    return;
}

const wip = columns.data.find(column => column.name === 'Work In Progress`);
if (!wip) {
    return;
}

const cards = await github.rest.projects.listCards({ 
    column_id: column.id, 
});
if (!cards) {
    return;
}

const card = cards.data.find(card => card.content_url === context.payload.issue.url);
if (!card) {
    return;
}

await github.rest.projects.moveCard({
    card_id: pr_card,
    position: 'bottom',
    column_id: wip.id,
}).catch(error => {})
```

File: .github/workflows/move-work-in-progress-prs.yml [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    types: [created]
+    branch:
+      - main
+
+jobs:
+  move-pr-to-wip:
+    if: github.event.comment.body == '/pr mark wip' && github.event.issue.pull_request
+    runs-on: ubuntu-latest
+    steps:
+      - name: Move To Initial Review
```
Comment (miscco) [inline .github/workflows/move-work-in-progress-prs.yml:16]:
```suggestion
      - name: Move To Work In Progress
```

File: .github/workflows/move-ready-for-review-prs.yml (hunk 3 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: It works, and the code doesn't look too bad, so this is unnecessary to look into now IMO

File: .github/workflows/move-ready-for-review-prs.yml (hunk 2 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: It is not possible, but the work being done is reduced greatly by the latest commit, so IMO no need to look into this.

File: .github/workflows/move-ready-for-review-prs.yml (hunk 1 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: Outdated place, will create new review with new locations

File: .github/workflows/move-ready-for-review-prs.yml (hunk 6 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: Done in latest commit

File: .github/workflows/move-ready-for-review-prs.yml (hunk 7 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: Yep, 99% what it looked like in the end.

File: .github/workflows/move-work-in-progress-prs.yml (hunk 8 above)
Comment (sam20908) [reply .github/workflows/move-work-in-progress-prs.yml]: Fixed

File: .github/workflows/move-ready-for-review-prs.yml (hunk 5 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: It was originally a typo, `work_in_progress` variable should've been `initial_review`.

File: .github/workflows/move-ready-for-review-prs.yml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            const pr_card = work_in_progress_cards.data.find(card => card.content_url === context.payload.issue.url)
+            if (!pr_card) {
+              return
+            }
+
+            github.rest.projects.moveCard({
+              card_id: pr_card.id,
+              position: 'bottom',
+              column_id: initial_review.id,
+            }).catch(error => {}) // FIXME figure out best way to handle error
```
Comment (sam20908) [inline .github/workflows/move-ready-for-review-prs.yml:59]: Reminder.

File: .github/workflows/move-work-in-progress-prs.yml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+                      github.rest.projects.moveCard({
+                        card_id: card.id,
+                        position: 'bottom',
+                        column_id: work_in_progress.id,
+                      }).catch(error => {}) // FIXME figure out best way to handle error
+                    }
+                  })
+                })
+                .catch(error => {}) // FIXME figure out best way to handle error
```
Comment (sam20908) [inline .github/workflows/move-work-in-progress-prs.yml:65]: Reminder.

File: .github/workflows/move-work-in-progress-prs.yml [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+              }).then(cards => {
+                  cards.data.forEach(card => {
+                    if (card.content_url === context.payload.issue.url) {
+                      // card is the PR card
+
+                      github.rest.projects.moveCard({
+                        card_id: card.id,
+                        position: 'bottom',
+                        column_id: work_in_progress.id,
+                      }).catch(error => {}) // FIXME figure out best way to handle error
```
Comment (sam20908) [inline .github/workflows/move-work-in-progress-prs.yml:61]: Reminder.

**Review by sam20908 — COMMENTED** [review]
New places of FIXME's.

**Comment by StephanTLavavej** [comment]
While the Code Reviews project has a Work In Progress column, we deleted the "work in progress" label a few months ago as it seemed to be redundant. We can re-add it at any time, though.

**Comment by sam20908** [comment]
This is officially ready for review, with everything addressed and the description updated.

File: .github/workflows/move-ready-for-review-prs.yml [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            // This avoids hardcoding the project ID
+            const projects = await github.rest.projects.listForRepo({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+            })
+            const code_reviews = projects.data.find(project => {
+              return project.name === 'Code Reviews';
+            })
+            if (!code_reviews) {
+              return
```
Comment (strega-nil-ms) [inline .github/workflows/move-ready-for-review-prs.yml:30]: It'd be nice to log if this happens (same below)

File: .github/workflows/move-work-in-progress-prs.yml [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            const work_in_progress = columns.data.find(column => column.name === 'Work In Progress')
+            if (!work_in_progress) {
+              console.error("'Work In Progress' column not found!")
+              return
+            }
+
+            columns.data.forEach(column => {
+              if (column.name !== 'Initial Review' && column.name !== 'Final Review') {
+                return // no reason to search through other columns and avoids unnecessary API calls
+              }
```
Comment (sam20908) [inline .github/workflows/move-work-in-progress-prs.yml:50]: I could add logging if neither of 'Initial Review' and 'Final Review' columns are not found. However, that would require extra variables to track that possibility just for the sake of logging. I am leaning towards not adding it to keep things simple, but feel free to object.

File: .github/workflows/move-ready-for-review-prs.yml (hunk 12 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml:31]: Done.

**Review by sam20908 — COMMENTED** [review]
One comment about the additional complexity to add logging for moving from 'Initial Review' or 'Final Review' column to 'Work in Progress' column.

File: .github/workflows/move-ready-for-review-prs.yml [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+          script: |
+            // Find "Code Reviews" project manually by name matching
+            // This avoids hardcoding the project ID
+            const projects = await github.rest.projects.listForRepo({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+            })
+            const code_reviews = projects.data.find(project => {
+              return project.name === 'Code Reviews';
+            })
```
Comment (strega-nil-ms) [inline .github/workflows/move-ready-for-review-prs.yml:28]:
```suggestion
            const code_reviews = projects.data.find(project => project.name === 'Code Reviews');
```

File: .github/workflows/move-ready-for-review-prs.yml [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+              return
+            }
+
+            const initial_review = columns.data.find(column => column.name === 'Initial Review')
+            if (!initial_review) {
+              console.error("'Initial Review' column not found!")
+              return
+            }
+
+            const work_in_progress_cards = await github.rest.projects.listCards({
```
Comment (strega-nil-ms) [inline .github/workflows/move-ready-for-review-prs.yml:53]: While it is true that the columns will probably never hit 30, we almost certainly need to implement pagination here.

Comment (strega-nil-ms) [reply .github/workflows/move-ready-for-review-prs.yml:53]: Something like:

```js
    const pr_card = github.paginate(github.rest.projects.listCards, {
      column_id: work_in_progress.id,
    }).then(cards => cards.find(card => card.content_url === context.payload.issue.url));
```

File: .github/workflows/move-ready-for-review-prs.yml [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+              console.error("Corresponding card for PR not found!")
+              return
+            }
+
+            github.rest.projects.moveCard({
+              card_id: pr_card.id,
+              position: 'bottom',
+              column_id: initial_review.id,
+            }).catch(error => {
+              console.error("Error occured while moving card to 'Initial Review'!")
```
Comment (strega-nil-ms) [inline .github/workflows/move-ready-for-review-prs.yml:67]:
```suggestion
              console.error(`Error occured while moving card to 'Initial Review': {error}`)
```

**Review by strega-nil-ms — COMMENTED** [review]
There're some clean ups possible

File: .github/workflows/move-ready-for-review-prs.yml [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+name: Move PR To Initial Review
+on:
+  issue_comment:
+    types: [created]
+    branch:
+      - main
+
+jobs:
+  move-pr-to-initial-review:
+    if: github.event.comment.body == '/pr mark ready' && github.event.issue.pull_request
```
[the remark this answers is not in the document]
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: Done

File: .github/workflows/move-ready-for-review-prs.yml (hunk 14 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: Done

File: .github/workflows/move-ready-for-review-prs.yml (hunk 16 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: Done

File: .github/workflows/move-ready-for-review-prs.yml (hunk 15 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml:51]: Yeah I did not think of that. I will transition all API calls necessary to use pagination.

Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: implemented

File: .github/workflows/move-ready-for-review-prs.yml [hunk 18]
```diff
@@ -64,5 +62,5 @@ jobs:
               position: 'bottom',
               column_id: initial_review.id,
             }).catch(error => {
-              console.error("Error occured while moving card to 'Initial Review'!")
+              console.error("Error occured while moving card to 'Initial Review': {error}")
```
Comment (strega-nil-ms) [inline .github/workflows/move-ready-for-review-prs.yml]: This won't do anything; you need to also switch the kind of quotes (to `` ` `` strings)

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

File: .github/workflows/move-ready-for-review-prs.yml (hunk 18 above)
Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: Crap I missed that

Comment (sam20908) [reply .github/workflows/move-ready-for-review-prs.yml]: Wait I did change that to backticks, I believe you are seeing outdated diff somehow...

Comment (strega-nil-ms) [reply .github/workflows/move-ready-for-review-prs.yml]: huh, odd.

**Review by strega-nil-ms — APPROVED** [review]

File: .github/workflows/move-work-in-progress-prs.yml [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+name: Move PR To Work In Progress
+on:
+  issue_comment:
+    types: [created]
+    branch:
+      - main
+
+jobs:
+  move-pr-to-wip:
+    if: contains(github.event.comment.body, '/pr mark wip') && github.event.issue.pull_request
```
Comment (StephanTLavavej) [inline .github/workflows/move-work-in-progress-prs.yml:13]: I observe that *anyone* can move a card, not just the PR author. While this is low risk, I think we can easily validate this. After some searching, I think that we want to compare `github.event.comment.user.login` to `github.event.issue.user.login`.

Comment (StephanTLavavej) [inline .github/workflows/move-work-in-progress-prs.yml:13]: Suggestion: I'm not sure if we get short-circuiting here, but it seems like a good idea in any event to phrase these conditions in order of increasing expense. For example, checking whether the "issue" is specifically a PR should be cheaper than doing a substring match, and it reads naturally.

Comment (StephanTLavavej) [inline .github/workflows/move-work-in-progress-prs.yml:13]: As we discussed on Discord, I'd like to make the commands a bit shorter, so people can remember them more easily: `/pr wip` and `/pr review`. (Similar to slash azp run.)

**Review by StephanTLavavej — COMMENTED** [review]
Thanks! I'll validate and push changes, one moment...

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I pushed a conflict-free merge with `main`, followed by minor typo fixes, semicolons, and adjusting the conditions to be short-circuit friendly, validating that the commenter is the same user as the PR author, and changing the commands to be easier to remember. I was unable to fully validate this in my repo since I don't have a classic project and can't make one anymore, but @CaseyCarter was able to help me validate the same-user check which is the part I was worried about (also this validated the YAML line wrapping I used; `>` drops newlines in the middle but keeps one at the end).

Note that the `if` condition is a [GitHub expression](https://docs.github.com/en/actions/learn-github-actions/expressions), so it uses a double equal, not the triple equal that's desirable in JS.

In the future, it may be desirable to fuse these two actions into one YAML file. Also, if we migrate to GitHub's modern projects, this code will need to change (no plans at the moment though).

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thank you for this major workflow improvement - this will be incredibly useful for contributors! :heart_eyes_cat: :100: :1st_place_medal:

**Pull request merged**